### PR TITLE
docs: added getting started with utility apis

### LINKED ISSUE #339: RFC: Feedback on API concept
Creating this issue to gather feedback on the API concept introduced in https://github.com/spotify/backstage/pull/251 and https://github.com/spotify/backstage/pull/326.

Anything goes, questions about purpose, ideas, clarifying things, naming, alternatives, etc.

Starting out with providing a few answers:

### Why was this created in the first place?

In the Spotify internal Backstage we added the concept as a way to make it easier to write tests, and to wrangle the growing complexity of the dependency graph between different app-wide features.

### What is the purpose of the API concept?

Apart from what's mentioned above, it's now also provides a way to decouple plugins from the APIs that they consume. A plugin can for example depend on an error reporting API, and it is up to the app that uses the plugin to provide an implementation of the API.

### Would every app have to write their own API implementations?

The `backstage/core`, or possibly a separate package, would ship a set of default implementations for all APIs. Apps can choose to use these directly, build upon them, or replace them.

### How does the API concept help with testing?

The `backstage/core` package would include mocks for all core API definitions. They should provide sane defaults for all behaviour and stay out of the way, but also provide additional methods for testing error/edge cases.

For example, any component consuming a core API in a way where we don't want to do any further testing of the API would simply do something like

```tsx
const rendered = render(wrapInTestApp(<MyComponent />))
```

While for example the error api mock could provide methods for ensuring that errors have been reported, something like

```tsx
const errorApi = testRegistry.get(errorApiRef) as MockErrorApi;

const rendered = render(wrapInTestApp(<MyComponent />))

expect(errorApi.reportedErrors.length).toBe(1);
```

### Are the API definitions limited only to the ones listed in backstage/core?

For now it would be the only supported way, but we want to support externally defined APIs, and APIs defined and consumed within a single plugin.

### PR DESCRIPTION
**Pull request by Rugvip**
Fixes #339

### CODE REVIEW

**Review by stefanalund — APPROVED** [review]

**Review by nikek — APPROVED** [review]
Looks good to me!

File: docs/getting-started/utility-apis.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# Utility APIs
+
+## Introduction
+
+Backstage Plugins strive to be self-contained, with as much functionality
+as possible residing withing the plugin itself and its backend APIs. There will however
```
Comment (stefanalund) [inline docs/getting-started/utility-apis.md:6]: withing -> within

File: docs/getting-started/utility-apis.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+always be a need for plugins to communicate outside of it's boundaries, both with other
```
Comment (freben) [inline docs/getting-started/utility-apis.md:7]: its

File: docs/getting-started/utility-apis.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  // Signal to the app that something went wrong, and display the error to the user.
+  const handleError = error => {
+    errorApi.post(error);
+  };
+
+  // the rest of the component ...
+};
+```
+
+Note that there is not explicit type given for the `ErrorApi`, this is because the `errorApiRef` has the type embedded, and `useApi` is able to infer the type.
```
Comment (freben) [inline docs/getting-started/utility-apis.md:42]: `is no explicit`

`for the ErrorApi. This is`

File: docs/getting-started/utility-apis.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+```ts
+builder.add(errorApiRef, new IgnoringErrorApi());
+```
+
+Note that the above line will cause an error if `IgnoreErrorApi` does not fully
+implement the `ErrorApi`, as it is checked by the type embedded in the `errorApiRef` at compile time.
+
+## Defining custom Utility APIs
+
+The pattern for plugins defining their own Utility APIs is not fully established yet. The current way is for the plugin to export it's own `ApiRef` and type for the API, along with one or more implementations. It is then up to the app to import, and register those APIs. See for example the lighthouse or graphiql plugins for examples of this.
```
Comment (freben) [inline docs/getting-started/utility-apis.md:104]: `export its own`

maybe make the words "lighthous" and "graphiql" link to the plugins, or even to where they export these apis

File: docs/getting-started/utility-apis.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+The goal is to make this process a bit smoother, but that requires work in other parts of Backstage, like configuration management. So it remains as a TODO. If you have more questions regarding this, or have an idea for and API that you want to share outside your plugin, hit us up in GitHub issues or the Backstage Discord server.
```
Comment (freben) [inline docs/getting-started/utility-apis.md:106]: `for an API`

Comment (freben) [reply docs/getting-started/utility-apis.md:106]: link to issues and discord?

File: docs/getting-started/utility-apis.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+## Architecture
+
+The `ApiRef` instances mentioned above provide a point of indirection between consumers and producers
```
Comment (freben) [inline docs/getting-started/utility-apis.md:110]: `producers of Utility`

File: docs/getting-started/utility-apis.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+The current method for connecting Utility API providers and consumers is via the React tree using
+an `ApiProvider`, which is added to the `AppProvider` of the `App`. In the future there may potentially
+be more ways to do this, in ways that are not tied to react. A design goal of the Utility APIs was to
+not have them directly tied to React.
+
+The indirection provided by Utility APIs also makes it straightforward to test components that depend
+on APIs, and to provide a standard common development environment for plugins. A proper test wrapper
+with mocked API implementations is not yet ready, but it will provided as a part of
+`@backstage/test-utils`. It will provide mocked variants of APIs, with additional methods for asserting
+a components interaction with the API.
```
Comment (freben) [inline docs/getting-started/utility-apis.md:134]: `a component's interaction`
